crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,567 @@ impl State {
+        if net_curve_trade == int256!(0) {
+            return Ok(share_proceeds);
+        }
+
+        // Proceed with Newton's method. The objective function, `F(x)`, is
+        // given by:
+        //
+        // F(x) = PV(x) * l - PV(0) * (l - w)
+        //
+        // Newton's method will terminate as soon as the current iteration is
+        // within the minimum tolerance or the maximum number of iterations has
+        // been reached.
+        let mut smallest_delta = I256::zero();
+        let mut closest_share_proceeds = fixed!(0);
+        let mut closest_present_value = fixed!(0);
+        for _ in 0..max(max_iterations, SHARE_PROCEEDS_MAX_ITERATIONS) {
+            // Clamp the share proceeds to the max share reserves delta since
+            // values above this threshold are always invalid.
+            share_proceeds = min(share_proceeds, max_share_reserves_delta);
+
+            // Simulate applying the share proceeds to the reserves.
+            //
+            // NOTE: We are calling this with 'self' so that original values are
+            // used with the updated value of share_proceeds
+            let updated_state =
+                self.get_state_after_liquidity_update(-I256::try_from(share_proceeds)?)?;
+
+            // Recalculate the present value.
+            let present_value = updated_state.calculate_present_value(current_block_timestamp)?;
+
+            // Short-circuit if we are within the minimum tolerance.
+            if self.should_short_circuit_distribute_excess_idle_share_proceeds(
+                active_lp_total_supply,
+                starting_present_value,
+                lp_total_supply,
+                present_value,
+            )? {
+                return Ok(share_proceeds);
+            }
+
+            // If the pool is net long, we can solve for the next iteration of
+            // Newton's method directly when the net curve trade is greater than
+            // or equal to the max bond amount.
+            if net_curve_trade > I256::zero() {
+                // Calculate the max bond amount.
+                let max_bond_amount = updated_state
+                    .calculate_max_sell_bonds_in_safe(self.minimum_share_reserves())?;
+
+                // If the net curve trade is greater than or equal to the max
+                // bond amount, we can solve directly for the share proceeds.
+                let net_curve_trade = FixedPoint::from(U256::try_from(net_curve_trade)?);
+                if net_curve_trade >= max_bond_amount {
+                    let share_proceeds = updated_state
+                        .calculate_distribute_excess_idle_share_proceeds_net_long_edge_case(
+                            self.share_adjustment(),
+                            self.share_reserves(),
+                            starting_present_value,
+                            active_lp_total_supply,
+                            withdrawal_shares_total_supply,
+                        )?;
```

Comment:
In Solidity, we do this:

```
                    if (!success) {
                        // NOTE: Return 0 to indicate that the share proceeds
                        // couldn't be calculated.
                        return 0;
                    }
```

I think it would be good to match this as closely as possible. This would help in teasing out which errors are unrecoverable and which aren't at the level of `calculate_distribute_excess_idle`.